Implement DHT event and Wallet FFI callback to be triggered when “enough” SAF responses are received

philipr-za opened the pull request:
## Description

This PR implements a DHT event bus with 1 new event that indicates that the DHT has received “enough” Store and Forward responses from its connected neighbours that we are relatively confident we have received any SAF responses due for us.

The SAF Handler on the inbound pipeline now sends a signal every time it handles a SAF response to the StoreAndForwardService. The StoreAndForwardService counts these signals and waits for the Event from the Connectivity manager to indicate when the node is considered Online. Online means that a proportion of neighbouring peers have been connected too. The SAF service then waits until it has received SAF response for at least the number of connected peers returned by the ConnectivityEvent and then it will fire the DhtEvent::StoreAndForwardMessagesReceived.

Currently a Node will not send a SAF response if it has no SAF messages to send but to enable this functionality this PR changes that logic to send an empty response so that the requesting node knows for sure there were no messages waiting for it.

The Wallet FFI Callback handler was extended to accept a new callback for to be called on this event. The Callback handler now subscribes to the DHT event bus and calls this new callback when the DhtEvent::StoreAndForwardMessagesReceived is received.

**NOTE: This PR will only start to work when the Peers in the network are also running this code because until then they will not respond if there are not SAF messages to be delivered.**

@kukabi and @Jasonvdb Please note the change to `wallet_create(…)`

## How Has This Been Tested?
Tests updated to test:
1. That the SAF middleware does signal when it receives a SAF response
2. That the SAF service does emit the event 
3. That the Callback handler does fire when it receives the event
4. That the SAF service does respond when it has no SAF messages to return.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

stringhandler reviewed (dismissed):
Looks good. Didn't test locally though

sdbondi commented on comms/dht/src/builder.rs:
```diff
@@ -75,6 +75,11 @@ impl DhtBuilder {
         self
     }
 
+    pub fn enable_auto_store_and_forward_requests(mut self) -> Self {
```
I've actually added something similar here: https://github.com/tari-project/tari/pull/2160/files#diff-91f3ccfba0c0a74bd4bac16bb43b8813 - might be better if use one (I vote for a single set function rather than enable/disable) - feel free to leave it as is and I'll clean it up when merging my PR, since it probably comes after this one

sdbondi commented on comms/dht/tests/dht.rs:
```diff
@@ -485,6 +500,7 @@ async fn dht_propagate_dedup() {
     let received = filter_received(collect_stream!(node_A_messaging, timeout = Duration::from_secs(20)));
     // Expected race condition: If A->(B|C)->(C|B) before A->(C|B) then (C|B)->A
     if received.len() > 0 {
+        println!("Events: {:?}", received);
```
```suggestion

```

philipr-za replied:
Woops

sdbondi commented on comms/dht/tests/dht.rs:
```diff
@@ -134,22 +142,23 @@ async fn setup_comms_dht(
-    .unwrap();
+    .with_num_neighbouring_nodes(8);
 
+    if with_saf {
```
Did it break any tests in this suite to have it always enabled?

If not - I say that you can just always have it on and avoid having an extra boolean param

philipr-za replied:
It did break a few tests that made assumptions on how many messages are sent and I was not comfortable updating them because I wasn't 100% sure what their intent was.

sdbondi replied:
Ah right, ye it makes sense to remove all the noise so that tests can remain focused

sdbondi replied:
Though now that I think about it - I think you should add an explicit call from node C to Node A (currently there is one for Node C -> B) - then I believe you wont need auto saf requests enabled

philipr-za replied:
Cool, done

sdbondi reviewed (commented):
LGTM - few inconsequential comments - going to test and approve if all goes well

philipr-za committed:
Implement DHT event and Wallet FFI callback for “enough” SAF responses

This PR implements a DHT event bus with 1 new event that indicates that the DHT has received “enough” Store and Forward responses from its connected neighbours that we are relatively confident we have received any SAF responses due for us.

The SAF Handler on the inbound pipeline now sends a signal every time it handles a SAF response to the StoreAndForwardService. The StoreAndForwardService counts these signals and waits for the Event from the Connectivity manager to indicate when the node is considered Online. Online means that a proportion of neighbouring peers have been connected too. The SAF service then waits until it has received SAF response for at least the number of connected peers returned by the ConnectivityEvent and then it will fire the DhtEvent::StoreAndForwardMessagesReceived.

Currently a Node will not send a SAF response if it has no SAF messages to send but to enable this functionality this PR changes that logic to send an empty response so that the requesting node knows for sure there were no messages waiting for it.

The Wallet FFI Callback handler was extended to accept a new callback for to be called on this event. The Callback handler now subscribes to the DHT event bus and calls this new callback when the DhtEvent::StoreAndForwardMessagesReceived is received.

**NOTE: This PR will only start to work when the Peers in the network are also running this code because until then they will not respond if there are not SAF messages to be delivered.**

@kukabi and @Jasonvdb Please note the change to `wallet_create(…)`

sdbondi reviewed (approved):
Ship it!
